akka-remote/src/main/scala/akka/remote/artery/Codecs.scala

```diff
@@ -247,19 +339,26 @@ private[remote] class Decoder(
   inboundContext:     InboundContext,
   system:             ExtendedActorSystem,
   uniqueLocalAddress: UniqueAddress,
-  compression:        InboundCompressions,
+  settings:           ArterySettings,
   bufferPool:         EnvelopeBufferPool,
-  inEnvelopePool:     ObjectPool[ReusableInboundEnvelope]) extends GraphStage[FlowShape[EnvelopeBuffer, InboundEnvelope]] {
+  inEnvelopePool:     ObjectPool[ReusableInboundEnvelope]) extends GraphStageWithMaterializedValue[FlowShape[EnvelopeBuffer, InboundEnvelope], InboundCompressionAccess] {
   import Decoder.Tick
   val in: Inlet[EnvelopeBuffer] = Inlet("Artery.Decoder.in")
   val out: Outlet[InboundEnvelope] = Outlet("Artery.Decoder.out")
   val shape: FlowShape[EnvelopeBuffer, InboundEnvelope] = FlowShape(in, out)
 
-  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new TimerGraphStageLogic(shape) with InHandler with OutHandler with StageLogging {
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, InboundCompressionAccess) = {
+    val logic = new TimerGraphStageLogic(shape) with InboundCompressionAccessImpl with InHandler with OutHandler with StageLogging {
       import Decoder.RetryResolveRemoteDeployedRecipient
+
+      override val compressions: InboundCompressions = {
+        if (settings.Advanced.Compression.Enabled) {
+          new InboundCompressionsImpl(system, inboundContext, settings.Advanced.Compression)
```

Comment:
hm, this requires some thinking. What happens if the inbound stream is restarted?
Then messages compressed with the old table may arrive, in best case we can't decompress them, in worse case it's a matching table version with different content and decompressed to wrong thing.

I'm not sure if this is a real problem.

If it is we need to create the instance in `ArteryTransport`.